component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -174,16 +186,18 @@ public boolean authenticateClient(HttpServletRequest request, Map<String, List>
     public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
                                    OAuthClientAuthnContext context) {
 
-        if (clientIdExistsAsParam(bodyParams) && validCertExistsAsAttribute(request)) {
+        String headerName = IdentityUtil.getProperty(MutualTLSUtil.MTLS_AUTH_HEADER);
+        if (clientIdExistsAsParam(bodyParams) && (validCertExistsAsAttribute(request) || (headerName != null &&
+                getCertificateFromHeader(request).isPresent()))) {
             if (log.isDebugEnabled()) {
-                log.debug("Client ID exists in request body parameters and a valid certificate found in request " +
-                        "attributes. Hence returning true.");
+                log.debug("Client ID exists in request body parameters and a valid certificate found in " +
+                        "the request. Hence returning true.");
             }
             return true;
         } else {
             if (log.isDebugEnabled()) {
-                log.debug("Mutual TLS authenticator cannot handle this request. Client id is not available in body " +
-                        "params or valid certificate not found in request attributes.");
+                log.debug("Mutual TLS authenticator cannot handle this request. " +
```

Comment:
IMO better to mention in the debug log from which method the certificate was found. That is either in request attributes or header..

Reply:
Fixed https://github.com/wso2-extensions/identity-oauth-addons/pull/54/commits/5e130533b27757d1c8cdf9424f5a9080bd6a7f4d